client/bqms_run/hooks.py

```diff
@@ -98,3 +103,25 @@ def un_expand(
             pattern, mapping, generator=expand, un_generator=un_expand
         )
     return MacroExpanderRouter(expanders)
+
+
+def preprocess_extract_ksh_heredoc_fragments(path: Path, text: str) -> str:
+    """
+    Extracts heredoc SQL fragments from any input ending in extension .ksh.
+    :param path: a bqms_run.paths.Path representing the relative path of the
+                 input to be preprocessed
+    :param text: a string representing the contents of the input to be
+                 preprocessed.
+    :return: for non-KSH inputs, the input text is returned unmodified. For
+                 KSH inputs, all heredoc SQL fragments in the input are
+                 concatenated together and returned. If an input contains
+                 no heredoc SQL fragments, a comment saying as much is returned.
+    """
+    if not path.match("*.ksh"):
+        return text
+    logger.info("Extracting heredoc fragments from KSH input: %s", path)
+    fragments = KshExtractor("bteq").read_fragments(text)
```

Comment:
In CompilerWorks, the command to use ("bteq" here) was configurable via Groovy DSL. I've left it as unconfigurable for now, unless and until we get a specific ask to do so, in which case we can add a config YAML option for it.

Reply:
(fyi)